### ISSUE TITLE
trouble packaging/uploading local files and directories

### ISSUE THREAD

**Issue by huemansdesign** [serverless/serverless-azure-functions#93]
Hi all, maybe Im not doing something correct. It seems like regardless of how I approach packaging local files/modules only my handler gets uploaded to Azure. I have to manually upload zip via Debug Console. I've tried packaging individually as well. Can you please tell me if I'm doing something wrong?

package:
  individually: true  
  exclude: 
    - libs/**
    - .bak/**

functions: 
  DService:
      handler: index.handler
      package:
        include:
          - config/**
          - DQuery.js 
          - index.js
          - modules/**
      events: 
        - http: true
          x-azure-settings:

**Comment by yaron-idan** [comment]
We also encountered this issue, and it seems to be solved in this PR - https://github.com/serverless/serverless-azure-functions/pull/83. Perhaps the maintainers should release a new version that will contain this changes.

**Comment by KosnIre** [comment]
Okay, so I wound up here because I had the exact problem in #34, but since this is the open ticket, I'll be describing my solution here. Sorry for the length.

At the time of writing, the most recent release is `serverless-azure-functions 0.4.0` and @yaron-idan is correct that #83 is **not** currently in a released version (it's just in `master`). So, here's what I did:

## Summary
1. `npm i --save git+ssh://git@github.com/serverless/serverless-azure-functions.git#b248b5bb514ad5e96d94204c957772171122a0b4`
2. `sls remove` and `sls deploy` (might not be necessary, but gets you a clean App)
3. a) Manually upload any `<yourFunction>-function.json` file via [https://<your-app-name>.scm.azurewebsites.net/dev](https://<your-app-name>.scm.azurewebsites.net/dev) in the corresponding function directory.
  b) Rename the uploaded file to `function.json` (you can right-click the file or hit F2)

## In-Depth

A couple of notes about my environment. I'm running `node v6.12.0` and `sls
My plugins look like this:
```
plugins:
  - serverless-webpack
  - serverless-offline
  - serverless-azure-functions
```
I'm also running `babel` and at the moment, both babel and webpack seem to be working properly.

1. `npm i --save git+ssh://git@github.com/serverless/serverless-azure-functions.git#b248b5bb514ad5e96d94204c957772171122a0b4`
That is the changeset that #83 got merged into `master`.

2. I also did a full `sls remove` and `sls deploy` here because Azure was giving me some errors (this might not be necessary for some people).

3. I still had a problem even after installing the fix and that is the Functions would not show up under the App in the Azure Portal. When I ran `sls invoke -f hello` I would get an error that looked like this:

```
Serverless: Invoke invoke
Serverless: Invoking function "hello"
Unhandled rejection Error
    at module.exports.logError.errorHandlingError (/usr/lib/node_modules/serverless/lib/classes/Error.js:94:11)
    at initializeErrorReporter.then.catch.e (/usr/lib/node_modules/serverless/bin/serverless:45:3)
    at runCallback (timers.js:672:20)
    at tryOnImmediate (timers.js:645:5)
    at processImmediate [as _immediateCallback] (timers.js:617:5)
From previous event:
    at __dirname (/usr/lib/node_modules/serverless/bin/serverless:43:9)
    at Object.<anonymous> (/usr/lib/node_modules/serverless/bin/serverless:46:4)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.runMain (module.js:604:10)
    at run (bootstrap_node.js:383:7)
    at startup (bootstrap_node.js:149:9)
    at bootstrap_node.js:496:3
```
This is happening because the plugin uploads the code and npm packages, but fails to upload the `function.json` file, which Azure uses to recognize the entry-points into the code. The solution is to manually upload the `function.json` file(s).

4. This seems unrelated to the issue, but in case you're a newbie to Azure Functions like me, I was able to `invoke` with 200 OK now, but still got no response text. For that, I had to edit my `serverless.yml` file and make sure I had 2 events for the function:
```
    events:
      - http: true  # 1st event has no 'name' as it defaults to 'res'
        x-azure-settings:
          route: hello
          authLevel: anonymous
      - http: true   # This 2nd event is important for getting responses
        x-azure-settings:
          direction: out
          name: res
```

**Comment by KosnIre** [comment]
Extra info about point 3) in my above comment. Maybe this will help the devs investigate. Or maybe, @Trane9991 would have an idea? For me, #83 does not upload a `function.json` file.

When I run `sls deploy` **_with_** the fix, my local package looks like this:
```
/<myService>/
--.serverless/
----hello.zip/
------node_modules/
------handler.js
------handler.js.map
------package.json
--.webpack/
--src/  # (other imported code in this dir)
--handler.js
--hello-function.json  # (this should be in the .zip I think?)
--serverless.yml
```

Even though my `sls deploy` logs show...
```
Serverless: Running Kudu command mv hello/hello-function.json hello/function.json...
Serverless: Syncing Triggers....Response statuscode: 200
Serverless: Successfully created Function App
```
...neither the `hello-function.json` nor the `function.json` file appears on the server.

**Comment by Trane9991** [comment]
@KosnIre do you have some "include/exclude" params in the *serverless.yml* file?

**Comment by Trane9991** [comment]
```
include:
- config/**
- DQuery.js
- index.js
- modules/**
```

Include probably will need `*function.json`

**Comment by KosnIre** [comment]
Thanks for the quick response. I tried using "include" several ways according to [the docs](https://serverless.com/framework/docs/providers/azure/guide/packaging/), but I can't get it to include _any extra_ file. So, I tried starting a new bare-minimum project and "include/exclude" works fine there. So, I think the difference is `serverless-webpack`. 

I found [a thread on their repo](https://github.com/serverless-heaven/serverless-webpack/issues/242#issuecomment-333356555) that talks about using `webpack-plugin-copy`. 

I was able to use that plugin to add the `hello-function.json` file to the .zip and when deployed, it was correctly renamed by the `Kudu` call. Now, the only problem is, it has to copy every file for every function. So in Azure, in every function directory, you'll have:
```
function.json
foo-function.json
bar-function.js
```
Not terrible, but I think there must be a way to make it copy only the correct file for each function. Maybe by writing an additional serverless plugin. I'll try playing with it more tomorrow.

**Comment by rupakg** [comment]
The title of the issue is misleading as I can deploy a basic app wherein the plugin will upload the function.json file. If this is a `webpack` related issue, lets change the title of the issue to reflect that.
@huemansdesign can you please confirm if you can deploy a basic app without `webpack`? Also, can you please format your `serverless.yml` file properly so that any yaml errors are evident?

**Comment by KosnIre** [comment]
@rupakg It's nice to see this issue getting some attention from a Member. I can understand how this might be confusing, so I'll try to help summarize and clear it up:

- **npm 0.4.0:** This code allows you to upload a function.json file and a single handler.js file... and that's it. But what if you want to `require()` code from other files? The require()'d files are not uploaded. Thus, your function will throw errors when invoked. This is why @huemansdesign said _"only my handler gets uploaded to Azure"_.
- **master:** This code (specifically the fixes from PR #83) allows you to upload a function.json file, a handler.js file, and any number of require()'d files (basically it zips your project folder and uploads that). This is fine, until you need to include a lot of `node_modules`, which slows down Azure Functions execution. So to fix that you need webpack...
- **PR #105:** This code makes `serverless-azure-functions` compatible with the `serverless-webpack` plugin, allowing you to package all of the `node_modules` into a single file, which Azure Functions can process much faster.

So, ultimately, there are 2 separate issues, but both issues involve getting more than 1 file (i.e. my project files & dependencies) uploaded to Azure. One issue is described here (how to do this _without webpack_), and one issue is described in #10 (how to do this _with webpack_).

I hope that helps :slightly_smiling_face:
